### PR TITLE
Add test coverage requirements for hellbender tools to README

### PR DESCRIPTION
**Pull request by droazen**
There have been requests for some additional clarity on "how
much test coverage is enough" for hellbender tools. Rather than
mandate a particular coverage target, I proposed a more flexible
set of guidelines which I've added to the README in this commit.

### CODE REVIEW

**Comment by droazen** [comment]
Please feel free to debate these requirements if you disagree!

**Comment by droazen** [comment]
@nh13 in particular should give feedback on whether these guidelines are useful, or if additional clarity is needed.

File: README.md [hunk 1]
```diff
@@ -32,6 +32,8 @@ General guidelines for Hellbender developers
 
 * Untested code is considered **non-existent** and thus is subject to removal at any time (exception is main methods, or super corner conditions, or `toString()` code).
 
+* Each tool should have at least one good end-to-end integration test with a check for expected output, plus high-quality unit tests for all non-trivial utility methods/classes used by the tool. Although we have no specific coverage target, coverage should be extensive enough that if tests pass, the tool is guaranteed to be in a usable state.
```
Comment (nh13) [inline README.md]: Guarantee is a strong word.  Perhaps we want that the author certifies it is in a usable state for its intended use?

Who defines the use cases for the integration or regression tests?  The tool developer?  The user of the tool?

Comment (droazen) [inline README.md]: How about "the tool is guaranteed to be in a usable state for its intended use(s)"? :)

For new tools, defining test cases is typically an iterative process: the initial set of test cases is proposed by the tool author in the initial pull request, then this set is potentially modified during code review if the reviewer notices significant cases that are missing. Additional test cases will then potentially be added post-merge as bug reports come in from the users.

For existing tools that lack tests, the process is much the same, except that the original tool author may not be available to write the initial set of test cases, so the task is more difficult. If the tool has an existing userbase, those who interact with the users (eg., the support group) could help devise a set of important use cases that should be covered by tests. If the tool has no userbase, we should probably debate whether it should remain in the codebase at all.

@akiezun @lbergelson please feel free to weigh in here as well

**Comment by lbergelson** [comment]
> Perhaps we want that the author certifies it is in a usable state for its intended use?
It's good that the author certifies it, but we are going to be tearing the tools apart and rebuilding them without necessarily understanding the details of the output.  We need to have tests that will tell us if we've broken them.

File: README.md (hunk 1 above)
Comment (droazen) [inline README.md]: @vdauwera should probably give feedback as well

**Comment by droazen** [comment]
Certainly the author can't just "certify" that the tool works without writing tests to prove it -- I don't think that was what @nh13 meant (but please correct me if I misunderstood). I read that as "if tests pass, the author certifies that the tool is in a usable state for its intended use(s)"

File: README.md (hunk 1 above)
Comment (vdauwera) [inline README.md]: Guys, @dekling has been going through the Picard tools defining minimal command lines for the basic use cases. If it helps, we can ask him to post the relevant info in each tool's test ticket so you guys can write the integration test at least. The second round of this project will be evaluating that the output is actually correct. If we find any problems at that stage we can revisit the corresponding tickets at that time.

Comment (nh13) [inline README.md]: @akiezun as the unit test requests were included in our sprint planning, can you please let me know ASAP if you want us to continue (or not)?

Comment (akiezun) [inline README.md]: @nh13 looking into what can be done with unit vs integration tests. will respond shortly

Comment (akiezun) [inline README.md]: @nh13 Yes, we need your help. I went through all classes with 0% coverage. Some of them can indeed be tested by simple integration tests. Testing most of those classes however requires a good understanding of what the algorithm does and multiple tests need to be written to account for all those cases. 

Below, all classes marked as 'non clp' must be unit tested. Only those marked as 'simple tests OK' are really possible to meaningfully test by a single input/output check. 

```
FixMateInformation	clp	
ReorderSam	clp	
FilterReads	clp	
RevertOriginalBaseQualitiesAndAddMateCigar	clp	
AddOrReplaceReadGroups	clp	
BuildBamIndex	clp	simple tests OK
DownsampleSam	clp	
ReplaceSamHeader	clp	simple tests OK
BamToBfq	clp	TO DELETE ?
SortSam	clp	simple tests OK
FilterReads.Filter	clp	
CalculateReadGroupChecksum	clp	simple tests OK
CompareSAMs	clp	simple tests OK
CheckTerminatorBlock	clp	
BamIndexStats	clp	simple tests OK
CollectWgsMetrics	clp	simple tests OK
CollectJumpingLibraryMetrics	clp	
TargetMetricsCollector	not clp	
HsMetricCollector	not clp	
CollectTargetedMetrics	not clp	
TargetedPcrMetricsCollector	not clp	
CalculateHsMetrics	clp	simple tests OK
CountingFilter	not clp	
CollectTargetedPcrMetrics	clp	simple tests OK
CountingMapQFilter	not clp	
CountingPairedFilter	not clp	
CountingDuplicateFilter	not clp	
TargetedPcrMetrics	class without code	nothing to test
HsMetrics	class without code	nothing to test
TargetMetrics	class without code	nothing to test
JumpingLibraryMetrics	class without code	nothing to test
CollectGcBiasMetrics	clp	
CollectOxoGMetrics	clp	
QualityScoreDistribution	clp	simple tests OK
CollectQualityYieldMetrics	clp	simple tests OK
MeanQualityByCycle	clp	simple tests OK
CollectBaseDistributionByCycle	clp	simple tests OK
CollectMultipleMetrics	clp	simple tests OK
CompareMetrics	clp	simple tests OK
GcBiasDetailMetrics	class without code	nothing to test
GcBiasSummaryMetrics	class without code	nothing to test
BaseDistributionByCycleMetrics	class without code	nothing to test
ClusterIntensityFileReader	not clp	
MultiTileParser	not clp	
FourChannelIntensityData	not clp	
IlluminaTextIterator	not clp	
MultiTileBclParser	not clp	
Casava18ReadNameEncoder	not clp	
IntensityChannel	class without code	nothing to test
MultiTileLocsParser	not clp	
MultiTileFilterParser	not clp	
IlluminaBasecallsToFastq	clp	
MarkIlluminaAdapters	clp	
GatherVcfs	clp	needs to test all cases 
LiftoverVcf	clp	simple tests OK - needs to test multiple cases
MergeVcfs	clp	
MakeSitesOnlyVcf	clp	simple tests OK but needs multiple cases
RenameSampleInVcf	clp	simple tests OK 
VcfToIntervalList	clp	simple tests OK
IntervalListTools	clp	
LiftOverIntervalList	clp	simple tests OK
EstimateLibraryComplexity	clp	
CollectHiSeqXPfFailMetrics	clp	
BclIndexReader	not clp	
GenotypeConcordanceStates	not clp	
ConvertSequencingArtifactToOxoG	clp	not sure
SequencingArtifactMetrics	not clp	
MemoryBasedReadEndsForMarkDuplicatesMap	not clp	
NormalizeFasta	clp	simple tests OK
ExtractSequences	clp	simple tests OK
```

Comment (nh13) [inline README.md]: I don't see any marked as ```non clp```. You will also need to define what simple tests you would like to see.  Basically, how will you accept the work?  Also, does this need to get done by the Engineering team or can you use other resources?

Comment (droazen) [inline README.md]: This list is super confusing -- instead of these cryptic abbreviations I'd recommend you edit it to assign one of the following values to each tool: "one integration test", "multiple integration tests", or "integration + unit tests", and for non-tool classes either "unit tests needed" or "no tests needed".

Also, do you agree with the requirement proposed above that each tool should have at a minimum one integration test?

Comment (akiezun) [inline README.md]: I agree that each tool should have at a minimum one integration test.

I'm going to merge this pull req. The discussion on what tools need what tests was moved to a spreadsheet for convenience. @nh13 and company will populate the issue list from that spreadsheet
